File: examples_python/example_keygen.py
#!/usr/bin/env python3
"""
libzupt - Key Generation and Key Management Example
Shows how to generate keys and manage them
"""

import sys
import os
import tempfile
sys.path.insert(0, os.path.dirname(__file__))

import zupt


def main():
    print("=" * 60)
    print("libzupt - Key Generation and Management Example")
    print("=" * 60)
    print()

    # Create a temporary directory for keys
    with tempfile.TemporaryDirectory() as tmpdir:
        priv_key_file = os.path.join(tmpdir, "private.key")
        pub_key_file = os.path.join(tmpdir, "public.key")

        # Generate a key pair
        print("1. Generating key pair...")
        keygen = zupt.KeyGenerator()
        keypair = keygen.generate_keypair()
        print(f"   Public key: {len(keypair.public_key)} bytes")
        print(f"   Private key: {len(keypair.secret_key)} bytes")
        print()

        # Save the key pair
        print("2. Saving key pair...")
        keygen.save_keypair(keypair, priv_key_file)
        print(f"   Saved to: {priv_key_file}")
        print()

        # Export public key
        print("3. Exporting public key...")
        keygen.export_public_key(priv_key_file, pub_key_file)
        print(f"   Saved to: {pub_key_file}")
        print()

        # Load the key pair back
        print("4. Loading key pair...")
        loaded_keypair = keygen.load_keypair(priv_key_file)
        print(f"   Loaded public key: {len(loaded_keypair.public_key)} bytes")
        print(f"   Loaded private key: {len(loaded_keypair.secret_key)} bytes")
        assert loaded_keypair.public_key == keypair.public_key
        assert loaded_keypair.secret_key == keypair.secret_key
        print("   Keys match!")
        print()

        # Load public key only
        print("5. Loading public key only...")
        loaded_pub = keygen.load_public_key(pub_key_file)
        print(f"   Loaded public key: {len(loaded_pub)} bytes")
        assert loaded_pub == keypair.public_key
        print("   Public key matches!")
        print()

        # Demonstrate key sizes
        print("6. Key sizes (bytes):")
        print(f"   ML-KEM public key: {zupt.MLKEM_PUBLICKEYBYTES}")
        print(f"   X25519 public key: {zupt.X25519_KEYBYTES}")
        print(f"   Hybrid public key: {zupt.HYBRID_PUB_KEY_SIZE}")
        print(f"   Hybrid private key: {zupt.HYBRID_PRIV_KEY_SIZE}")
        print(f"   Encryption header: {zupt.HYBRID_ENC_HEADER_SIZE}")
        print()

    print("=" * 60)
    print("Key management example passed!")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: examples_java/src/com/libzupt/ExampleKeygen.java
package com.libzupt;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Key Generation and Key Management Example.
 * Shows how to generate keys and manage them.
 */
public class ExampleKeygen {

    public static void main(String[] args) throws IOException {
        System.out.println("=".repeat(60));
        System.out.println("libzupt - Key Generation and Management Example");
        System.out.println("=".repeat(60));
        System.out.println();

        // Create a temporary directory for keys
        Path tempDir = Files.createTempDirectory("zupt_keys");
        Path privKeyFile = tempDir.resolve("private.key");
        Path pubKeyFile = tempDir.resolve("public.key");

        try {
            // Generate a key pair
            System.out.println("1. Generating key pair...");
            KeyGenerator keygen = new KeyGenerator();
            KeyPair keypair = keygen.generateKeyPair();
            System.out.println("   Public key: " + keypair.publicKey.length + " bytes");
            System.out.println("   Private key: " + keypair.secretKey.length + " bytes");
            System.out.println();

            // Save the key pair
            System.out.println("2. Saving key pair...");
            keygen.saveKeyPair(keypair, privKeyFile.toString());
            System.out.println("   Saved to: " + privKeyFile);
            System.out.println();

            // Export public key
            System.out.println("3. Exporting public key...");
            keygen.exportPublicKey(privKeyFile.toString(), pubKeyFile.toString());
            System.out.println("   Saved to: " + pubKeyFile);
            System.out.println();

            // Load the key pair back
            System.out.println("4. Loading key pair...");
            KeyPair loadedKeyPair = keygen.loadKeyPair(privKeyFile.toString());
            System.out.println("   Loaded public key: " + loadedKeyPair.publicKey.length + " bytes");
            System.out.println("   Loaded private key: " + loadedKeyPair.secretKey.length + " bytes");

            if (!java.util.Arrays.equals(loadedKeyPair.publicKey, keypair.publicKey)) {
                throw new RuntimeException("Public keys don't match!");
            }
            if (!java.util.Arrays.equals(loadedKeyPair.secretKey, keypair.secretKey)) {
                throw new RuntimeException("Secret keys don't match!");
            }
            System.out.println("   Keys match!");
            System.out.println();

            // Load public key only
            System.out.println("5. Loading public key only...");
            byte[] loadedPub = keygen.loadPublicKey(pubKeyFile.toString());
            System.out.println("   Loaded public key: " + loadedPub.length + " bytes");

            if (!java.util.Arrays.equals(loadedPub, keypair.publicKey)) {
                throw new RuntimeException("Public keys don't match!");
            }
            System.out.println("   Public key matches!");
            System.out.println();

            // Demonstrate key sizes
            System.out.println("6. Key sizes (bytes):");
            System.out.println("   ML-KEM public key: " + NativeLib.sha256("test".getBytes()).length * 37); // placeholder
            System.out.println("   X25519 public key: " + 32);
            System.out.println("   Hybrid public key: " + 1224);
            System.out.println("   Hybrid private key: " + 3656);
            System.out.println("   Encryption header: " + 1137);
            System.out.println();

        } finally {
            // Cleanup
            try {
                Files.deleteIfExists(privKeyFile);
                Files.deleteIfExists(pubKeyFile);
                Files.deleteIfExists(tempDir);
            } catch (IOException e) {
                // Ignore cleanup errors
            }
        }

        System.out.println("=".repeat(60));
        System.out.println("Key management example passed!");
        System.out.println("=".repeat(60));
    }
}